File: LookingForAJob/MyApplication_2025/app/src/main/java/com/example/myapplication/okhttp/DownloadFileDemo.java
package com.example.myapplication.okhttp;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;

/**
 * 下载单个文件
 */
public class DownloadFileDemo {
    private OkHttpClient client;

    public DownloadFileDemo() {
        client = new OkHttpClient();
    }

    public void downloadFileAsync(String fileUrl, String destinationPath) {
        Request request = new Request.Builder()
                .url(fileUrl)
                .build();

        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                System.err.println("下载失败: " + e.getMessage());
            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                if (!response.isSuccessful()) {
                    throw new IOException("服务器返回错误: " + response.code());
                }

                ResponseBody body = response.body();
                if (body == null) {
                    throw new IOException("响应体为空");
                }

                File outputFile = new File(destinationPath);
                try (InputStream inputStream = body.byteStream();
                     FileOutputStream outputStream = new FileOutputStream(outputFile)) {
                    byte[] buffer = new byte[4096];
                    int bytesRead;
                    while ((bytesRead = inputStream.read(buffer)) != -1) {
                        outputStream.write(buffer, 0, bytesRead);
                    }
                    System.out.println("文件下载成功，保存到: " + destinationPath);
                } catch (IOException e) {
                    System.err.println("文件写入失败: " + e.getMessage());
                }
            }
        });
    }

    public void main(String[] args) {

        String fileUrl = "https://example.com/path/to/file.zip";
        String destinationPath = "path/to/downloaded/file.zip";
        downloadFileAsync(fileUrl, destinationPath);
    }
}

File: LookingForAJob/MyApplication_2025/app/src/main/java/com/example/myapplication/retrofit/DownloadFileDemo.kt
package com.example.myapplication.retrofit

import android.util.Log
import okhttp3.OkHttpClient
import okhttp3.ResponseBody
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

/**
 * 下载单个文件
 */
class DownloadFileDemo {

    companion object RetrofitClient {
        private const val BASE_URL = "https://example.com/"

        private val client: OkHttpClient = OkHttpClient.Builder()
            .addInterceptor(HttpLoggingInterceptor().apply {
                level = HttpLoggingInterceptor.Level.BODY
            })
            .build()

        val instance: Retrofit = Retrofit.Builder()
            .baseUrl(BASE_URL)
            .client(client)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
    }

    fun test(){
        val fileUrl = "https://example.com/file.zip"
        downloadFile(fileUrl)
    }

    private fun downloadFile(fileUrl: String) {
        val service = RetrofitClient.instance.create(ApiService::class.java)
        val call = service.downloadFile(fileUrl)

        call.enqueue(object : Callback<ResponseBody> {
            override fun onResponse(call: Call<ResponseBody>, response: Response<ResponseBody>) {
                if (response.isSuccessful && response.body() != null) {
                    saveFile(response.body()!!)
                } else {
                    Log.e("FileDownload", "Failed to download file: ${response.errorBody()}")
                }
            }

            override fun onFailure(call: Call<ResponseBody>, t: Throwable) {
                Log.e("FileDownload", "Error: ${t.message}")
            }
        })
    }

    private fun saveFile(responseBody: ResponseBody) {
        val filesDir = ""
        //指定文件保存路径
        val file = File(filesDir, "downloaded_file.zip")
        try {
            //在ResponseBody对象中获取输入流
            val inputStream = responseBody.byteStream()
            val outputStream = FileOutputStream(file)
            // 4KB buffer
            val buffer = ByteArray(4 * 1024)
            var read: Int
            //通过输出流把文件保存到设备
            while (inputStream.read(buffer).also { read = it } != -1) {
                outputStream.write(buffer, 0, read)
            }
            outputStream.flush()
            outputStream.close()
            inputStream.close()
            Log.d("FileDownload", "File saved to ${file.absolutePath}")
        } catch (e: IOException) {
            Log.e("FileDownload", "Failed to save file: ${e.message}")
        }
    }
}